include/picongpu/particles/collision/InterCollision.hpp

```diff
@@ -29,480 +29,427 @@
+            float_X const densityLongList = maxListLength == size0 ? density0 : density1;
+            float_X const densityShortList = densityLongList == density1 ? density0 : density1;
 
-                    if(sizeShort != 0u)
-                        for(uint32_t i = 0; i < sizeLong; ++i)
-                        {
-                            auto parLong = detail::getParticle(pBoxLong, frameLong, listLong[i]);
-                            auto parShort = detail::getParticle(pBoxShort, frameShort, listShort[i % sizeShort]);
-                            destCollisionFunctor.duplicationCorrection = duplicationCorrection(i, sizeShort, sizeLong);
-                            destCollisionFunctor(detail::makeCollisionContext(worker, rngHandle), parLong, parShort);
-                        }
+            auto destCollisionFunctor
+                = srcCollisionFunctor(worker, localSuperCellOffset, densityLongList, densityShortList, maxListLength);
 
-                    return destCollisionFunctor;
-                }
-            };
+            if constexpr(useScreeningLength)
+            {
+                auto const shifted = screeningLengthSquared.shift(superCellIdx * SuperCellSize::toRT());
+                auto const idxInSuperCell = pmacc::math::mapToND(SuperCellSize::toRT(), linearCellIdx);
+                destCollisionFunctor.coulombLogFunctor.screeningLengthSquared_m = shifted(idxInSuperCell)[0];
+            }
 
-            /* Run kernel for collisions between two species.
-             *
-             * @tparam T_CollisionFunctor A binary particle functor defining a single macro particle collision in
-             * the binary-collision algorithm.
-             * @tparam T_FilterPair A pair of particle filters, each for each species
-             *     in the colliding pair.
-             * @tparam T_Species0 1st colliding species.
-             * @tparam T_Species1 2nd colliding species.
-             */
-            template<
-                typename T_CollisionFunctor,
-                typename T_FilterPair,
-                typename T_Species0,
-                typename T_Species1,
-                uint32_t colliderId,
-                uint32_t pairId>
-            struct DoInterCollision
+            if(minListLength != 0u)
             {
-                /* Run kernel
-                 *
-                 * @param deviceHeap A pointer to device heap for allocating particle lists.
-                 * @param currentStep The current simulation step.
-                 */
-                HINLINE void operator()(const std::shared_ptr<DeviceHeap>& deviceHeap, uint32_t currentStep)
+                uint32_t const numDupParticles = maxListLength - minListLength;
+                for(uint32_t i = 0; i < numDupParticles; ++i)
+                {
+                    auto par0 = parAccessor0[i % size0];
+                    auto par1 = parAccessor1[i % size1];
+                    destCollisionFunctor.duplicationCorrection
```

Comment:
@pordyna We need to talk about this duplication correction to take care that I have not broken anything.